core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -47,12 +47,35 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
 
   private val transportConf = SparkTransportConf.fromSparkConf(conf)
 
-  def getDataFile(shuffleId: Int, mapId: Int): File = {
-    blockManager.diskBlockManager.getFile(ShuffleDataBlockId(shuffleId, mapId, NOOP_REDUCE_ID))
+  private def getDataFile(
+      shuffleId: Int,
+      mapId: Int,
+      blockManagerId: BlockManagerId = blockManager.blockManagerId)
+    : File = {
+    if (blockManager.blockManagerId != blockManagerId) {
+      blockManager.diskBlockManager.getShuffleFileBypassNetworkAccess(
+        ShuffleDataBlockId(shuffleId, mapId, NOOP_REDUCE_ID), blockManagerId)
+    } else {
+      blockManager.diskBlockManager.getFile(
+        ShuffleDataBlockId(shuffleId, mapId, NOOP_REDUCE_ID))
+    }
   }
 
-  private def getIndexFile(shuffleId: Int, mapId: Int): File = {
-    blockManager.diskBlockManager.getFile(ShuffleIndexBlockId(shuffleId, mapId, NOOP_REDUCE_ID))
+  def getDataFile(shuffleId: Int, mapId: Int): File =
+    getDataFile(shuffleId, mapId, blockManager.blockManagerId)
+
+  private def getIndexFile(
+      shuffleId: Int,
+      mapId: Int,
+      blockManagerId: BlockManagerId = blockManager.blockManagerId)
+    : File = {
+    if (blockManager.blockManagerId != blockManagerId) {
+      blockManager.diskBlockManager.getShuffleFileBypassNetworkAccess(
+        ShuffleIndexBlockId(shuffleId, mapId, NOOP_REDUCE_ID), blockManagerId)
+    } else {
+      blockManager.diskBlockManager.getFile(
+        ShuffleIndexBlockId(shuffleId, mapId, NOOP_REDUCE_ID))
```

Comment:
actually, if you get rid of `getShuffleFileBypassNetworkAccess` as I suggested elsewhere, then you don't need to abstract this since it's just 1 line.